nemo/collections/asr/parts/preprocessing/features.py

```diff
@@ -428,5 +476,175 @@ def forward(self, x, seq_len):
+            )
+
+            # log features if required
+            if self.log:
+                eps = self.log_zero_guard_value_fn(torch.float32)
+                if self.log_zero_guard_type == "add":
+                    spec = spec + eps
+                elif self.log_zero_guard_type == "clamp":
+                    spec = dali.math.max(spec, eps)
+                else:
+                    raise ValueError("log_zero_guard_type was not understood")
+                # Natural Logarithm
+                spec = dali.fn.to_decibels(spec, multiplier=math.log(10), reference=1.0, cutoff_db=-120)
+
+            if self.frame_splicing > 1:
+                spec = self._daligraph_splice_frames(spec, self.nfilt, spec_len, stacking=self.frame_splicing)
+
+            # Trimming Spectrogram to match the reference implementation
+            start = dali.types.Constant(0, shape=[], dtype=dali.types.INT64, device='cpu')
+            spec = dali.fn.slice(spec, start, spec_len, axes=(1,))
+
+            # Normalization
+            normalization_axes = None
+            if self.normalize:
+                if self.normalize == "per_feature":
+                    normalization_axes = [1]
+                elif self.normalize == "all_features":
+                    normalization_axes = [0, 1]
+                elif "fixed_mean" in self.normalize and "fixed_std" in self.normalize:
+                    raise ValueError("Normalization with fixed mean/stddev not yet supported.")
+                    # TODO: implement
+                else:
+                    raise ValueError(f"Unknown normalization type: {self.normalize}")
+
+                # Normalization
+                spec = dali.fn.normalize(spec, axes=normalization_axes, epsilon=1e-5 ** 2, ddof=1)
+
+            # Pads temporal dimension to take the length of the longest sample in the batch, padded up to a multiple of ``pad_to``
+            pad_align = (self.pad_to,) if self.pad_to > 0 else None
+            pad_shape = (self.max_length,) if self.pad_to == 'max' else (-1,)
+            spec = dali.fn.pad(spec, fill_value=self.pad_value, axes=(1,), align=pad_align, shape=pad_shape)
+            if device == 'gpu':
+                spec_len = spec_len.gpu()
+        pipe.set_outputs(spec, spec_len)
+        # Building DALI pipeline
+        pipe.build()
+        return pipe
+
+    def forward_dali(self, x, x_len):
+        device = x.device
+        device_str = str(device)
+        device_str_toks = device_str.split(':')
+        device_type_str = 'gpu' if device_str_toks[0] == 'cuda' else 'cpu'
+        device_id = None
+        if device_type_str == 'gpu':
+            device_id = int(device_str_toks[1])
+            cuda_stream = torch.cuda.current_stream(device=device)
+        else:
+            device_id = None
+            cuda_stream = None
+        batch_size = x.shape[0]
+        if (
+            self.dali_ctx.batch_size != batch_size
```

Comment:
So this will rebuild the entire pipeline each time batch size changes? Under DDP this should probably be fine, for single GPU might cause a bit of slowdown at the end of each training epoch?

Reply:
I am wondering, why would the batch size change at the end of an epoch?

Reply:
You are right. I think we can make it rebuild only if the batch size increases. I'll try that.